src/linux.rs

```diff
@@ -0,0 +1,398 @@
+
+    let cpus = CGROUPS_CPUS.load(Ordering::Acquire);
+
+    if cpus > 0 {
+        Some(cpus)
+    } else {
+        None
+    }
+}
+
+fn init_cgroups() {
+    // Should only be called once
+    debug_assert!(CGROUPS_CPUS.load(Ordering::SeqCst) == 0);
+
+    match load_cgroups("/proc/self/cgroup", "/proc/self/mountinfo") {
+        Some(quota) => {
+            if quota == 0 {
+                return;
+            }
+
+            let logical = logical_cpus();
+            let count = ::std::cmp::min(quota, logical);
+
+            CGROUPS_CPUS.store(count, Ordering::SeqCst);
+        }
+        None => return,
+    }
+}
+
+fn load_cgroups<P1, P2>(cgroup_proc: P1, mountinfo_proc: P2) -> Option<usize>
+where
+    P1: AsRef<Path>,
+    P2: AsRef<Path>,
+{
+    let subsys = some!(Subsys::load_cpu(cgroup_proc));
+    let mntinfo = some!(MountInfo::load_cpu(mountinfo_proc));
+    let cgroup = some!(Cgroup::mount(mntinfo, subsys));
+    cgroup.cpu_quota()
+}
+
+struct Cgroup {
+    base: PathBuf,
+}
+
+struct MountInfo {
+    root: String,
+    mount_point: String,
+}
+
+struct Subsys {
+    base: String,
+}
+
+impl Cgroup {
+    fn new(dir: PathBuf) -> Cgroup {
+        Cgroup {
+            base: dir,
+        }
+    }
+
+    fn mount(mntinfo: MountInfo, subsys: Subsys) -> Option<Cgroup> {
+        // Translate the subsystem directory via the host paths.
+
+        debug!("subsys = {:?}; root = {:?}; mount_point = {:?}", subsys.base, mntinfo.root, mntinfo.mount_point);
+
+        let rel_from_root = some!(Path::new(&subsys.base).strip_prefix(&mntinfo.root).ok());
+
+        debug!("rel_from_root: {:?}", rel_from_root);
+
+        // join(mp.MountPoint, relPath)
+        let mut path = PathBuf::from(mntinfo.mount_point);
+        path.push(rel_from_root);
+        Some(Cgroup::new(path))
+    }
+
+    fn cpu_quota(&self) -> Option<usize> {
+        let quota_us = some!(self.quota_us());
+        let period_us = some!(self.period_us());
+
+        // protect against dividing by zero
+        if period_us == 0 {
+            return None;
+        }
+
+        Some(quota_us / period_us)
+    }
+
+    fn quota_us(&self) -> Option<usize> {
+        self.param("cpu.cfs_quota_us")
+    }
+
+    fn period_us(&self) -> Option<usize> {
+        self.param("cpu.cfs_period_us")
+    }
+
+    fn param(&self, param: &str) -> Option<usize> {
+        let mut file = some!(File::open(self.base.join(param)).ok());
+
+        let mut buf = String::new();
+        some!(file.read_to_string(&mut buf).ok());
+
+        buf.trim().parse().ok()
+    }
+}
+
+impl MountInfo {
+    fn load_cpu<P: AsRef<Path>>(proc: P) -> Option<MountInfo> {
+        let file = some!(File::open(proc).ok());
+        let file = BufReader::new(file);
+
+        file.lines()
+            .filter_map(|result| result.ok())
+            .filter_map(MountInfo::parse_line)
+            .next()
+    }
+
+    fn parse_line(line: String) -> Option<MountInfo> {
+        let mut fields = line.split(' ');
```

Comment:
This basically assumes the paths never contain any spaces, is this a safe assumption to make?

Reply:
In general no, but this will just skip (as in, returning `None`) any lines that it fails to match due to spaces embedded in paths. I've never seen a cgroup hierarchy mounted on a space-containing path, so I think this won't fail in practice.

Alternatively, `/proc/self/mounts` may be slightly easier to parse while being space-correct: the `cgroup` type can be directly detected as the first token, the path is enclosed between type-tokens, and there are overall less entries to track.

Reply:
FWIW, this technique for parsing the files is the same used in JDK and in automaxprocs.go.